### PR TITLE
Add C# coding standard #21

### LINKED ISSUE #21: Add C# coding standard

### PR DESCRIPTION
**Pull request by AngShiYa**
Fixes #21 
Following the standards from [CSharpGuidelines](https://github.com/dennisdoomen/CSharpGuidelines).
Some examples were taken from other sources listed under references.
[Preview](https://github.com/AngShiYa/process/blob/add-csharp-coding-standard/codingStandards/CodingStandard-CSharp.md)

### CODE REVIEW

**Comment by acjh** [comment]
While you wait for a review, take note of the following guidelines:
1. Include a `[Preview](https://github.com/AngShiYa/process/blob/add-csharp-coding-standard/codingStandards/CodingStandard-CSharp.md)` in your PR description.
2. Follow the [naming convention for branch names](https://github.com/oss-generic/process/blob/master/docs/FormatsAndConventions.md#branch) for subsequent PRs.

**Comment by AngShiYa** [comment]
Noted and updated the description.

File: codingStandards/CodingStandard-CSharp.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+![](Good.png)
+
+```cs
+var startInfo = new ProcessStartInfo("myapp.exe");
+{
+	StandardOutput = Console.Output,
+	UseShellExecute = true
+};
+
```
Comment (acjh) [inline codingStandards/CodingStandard-CSharp.md]: Remove unnecessary newline.

**Comment by AngShiYa** [comment]
@acjh Updated.

File: codingStandards/CodingStandard-CSharp.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+var startInfo = new ProcessStartInfo("myapp.exe");
+startInfo.StandardOutput = Console.Output;
+startInfo.UseShellExecute = true;
+```
+
+## References
+1. [C# Coding Coventions (C# Programming Guide)](https://msdn.microsoft.com/en-us/library/ff926074.aspx)<br>
+2. [C# Coding Standards for .NET](http://se.inf.ethz.ch/old/teaching/ss2007/251-0290-00/project/CSharpCodingStandards.pdf)<br>
+3. [CSharpGuidelines](https://github.com/dennisdoomen/CSharpGuidelines)<br>
+4. [C# Coding Style Guide](http://www.icsharpcode.net/TechNotes/SharpDevelopCodingStyle03.pdf)
```
Comment (acjh) [inline codingStandards/CodingStandard-CSharp.md]: Swop the links of `3.` and `4.` for alphabetical order.

Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Based on our markdown coding standard, its preferred if you use `1.` for each item. That way, it is easier to insert items later.

File: codingStandards/CodingStandard-CSharp.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
+# C# Coding Standard
+
+**Important:** Use [CSharpGuidelines](https://github.com/dennisdoomen/CSharpGuidelines) for any topics not covered in this document.
```
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Any reason to choose this one over the one from Microsoft? Is this one different from the Microsoft one?

File: codingStandards/CodingStandard-CSharp.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+**3. Maintain a common order for each file.**<br>
+i. Member variable<br>
+ii. Constructors and Finalizer<br>
+iii. Nested Enums, Structs and Classes<br>
+iv. Properties<br>
+v. Methods<br>
+> Sequence declaration within type groups are based on visibility in this order: `public, protected, internal, private`.
+
+### Indentation
+**1. Basic indentation should be 4 spaces.**<br>
```
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Also add `(not tabs)`?

File: codingStandards/CodingStandard-CSharp.md [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+**1. Source file name and class name should always match.**<br>
+MyClass.cs → `public class MyClass { ... }`
+
+**2. Directory names should follow namespace for the class.**<br>
+`System.Windows.Forms.Control` should use the path System\Windows\Forms\Control.cs
+> Do not use namespace name with dots. This will make it easier to map namespaces to the directory layout.
+
+## Formatting
+### Layout
+**1. Avoid putting multiple namespaces or classes per file.**<br>
```
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: If you use headings (e.g. `####`), GitHub auto-generates anchors that can be used from other places to refer to a specific rule easily.

File: codingStandards/CodingStandard-CSharp.md (hunk 3 above)
Comment (AngShiYa) [inline codingStandards/CodingStandard-CSharp.md]: I feel that the CSharpGuidelines is more comprehensive for beginners. It includes basic stuff like casing convention, layout while the Microsoft one doesn't.

**Comment by AngShiYa** [comment]
@damithc Updated based on your comments.

File: codingStandards/CodingStandard-CSharp.md [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+![](Bad.png)
+
+```cs
+var startInfo = new ProcessStartInfo("myapp.exe");
+startInfo.StandardOutput = Console.Output;
+startInfo.UseShellExecute = true;
+```
+
+## References
+1. [C# Coding Coventions (C# Programming Guide)](https://msdn.microsoft.com/en-us/library/ff926074.aspx)<br>
```
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Perhaps we can give some info about each link? e.g.

1. [C# Coding Coventions (C# Programming Guide)](https://msdn.microsoft.com/en-us/library/ff926074.aspx) -- From Microsoft<br>

File: codingStandards/CodingStandard-CSharp.md (hunk 3 above)
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Add some info about how you constructed this standard. e.g.

> The coding standard is primarily based on [CSharpGuidelines](https://github.com/dennisdoomen/CSharpGuidelines)
> (by John Doe) but differs from it in these ways:
>
> * ...
>  Some examples were taken from ...

**Comment by AngShiYa** [comment]
@damithc Updated.

File: codingStandards/CodingStandard-CSharp.md [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+![](Bad.png)
+
+```cs
+var startInfo = new ProcessStartInfo("myapp.exe");
+startInfo.StandardOutput = Console.Output;
+startInfo.UseShellExecute = true;
+```
+
+## References
+1. [C# Coding Coventions (C# Programming Guide)](https://msdn.microsoft.com/en-us/library/ff926074.aspx) -- From Microsoft<br>
```
Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: `<br>` needed?

Comment (damithc) [inline codingStandards/CodingStandard-CSharp.md]: Add blank line at the beginning of a list

**Comment by AngShiYa** [comment]
@damithc Updated.

**Pull request merged**